### ISSUE TITLE
Inconsistent result order in Notebooks

### ISSUE THREAD

**Issue by cowhitley** [microsoft/azuredatastudio#24095]
Type: <b>Bug</b>

Run the following Kusto Query multiple times to see different results from the Kusto Detective Agency Season 2 Onboarding.

DetectiveCases
| where datetime_part('Year', Timestamp) == 2022
| where EventType == 'CaseOpened'
| project CaseOpened = Timestamp, CaseId, Properties
| extend Bounty = toreal(Properties.Bounty)
| join kind=inner
    (
    DetectiveCases
    | where EventType == 'CaseSolved'
    | summarize arg_min(Timestamp, DetectiveId) by CaseId)
    on CaseId
| summarize BountyEarned = sum(Bounty) by DetectiveId
| sort by BountyEarned desc

Extension version: 0.5.8
Azure Data Studio version: azuredatastudio 1.45.0 (8eff468100f02157f2219335f7a566d14451ca49, 2023-07-24T18:20:47.177Z)
OS version: Darwin arm64 22.5.0
Restricted Mode: No
Preview Features: Disabled
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Apple M2 Max (12 x 24)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled|
|Load (avg)|2, 2, 2|
|Memory (System)|96.00GB (52.04GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|
</details>

**Comment by Charles-Gagnon** [comment]
@alanrenmsft @lucyzhang929 @chlafreniere This seems to be a regression caused by https://github.com/microsoft/azuredatastudio/pull/23378 - although that's not really the "root" problem here.

The main problem I see is that as far as I can tell we have no ordering guarantees in the result streaming logic. We start off by getting a number of result set updates here : 

https://github.com/microsoft/azuredatastudio/blob/f750de94b7c938ce023e6ff3ebc5027821efb1a6/src/sql/workbench/services/notebook/browser/sql/sqlSessionManager.ts#L572

These come in one at a time in order, but are async so aren't blocking. Each result set summary in this case contains a subset of the overall result set (streaming data).

Each of these then call queryAndConvert data to get the actual rows for that result set.

https://github.com/microsoft/azuredatastudio/blob/f750de94b7c938ce023e6ff3ebc5027821efb1a6/src/sql/workbench/services/notebook/browser/sql/sqlSessionManager.ts#L605

These call into queryRunner.getQueryRows. Previously, each of these calls would result in a single RPC call to the tools service to get the row data. This happened to work out because the tools service only handles one request at a time - so it would sort of guarantee that the results would come back in order. This would look like this

Request RS1 rows
Request RS2 rows (blocked on RS1 message)
Request RS3 rows (blocked on RS1 + RS2 message)
Get RS1 rows (starts processing RS2 message)
Get RS2 rows (starts processing RS3 message)
Get RS3 rows

But in the above PR this call was split into multiple "pages" - so each result set could result in multiple separate calls to the tools service. Because all of these calls from the multiple result sets are async this means we could have the result set calls be interweaved - so you could end up with something like this

Request 500 rows from RS1
Request 500 rows from RS2 (blocked on RS1 message)
Request 500 rows from RS3 (blocked on RS1 + RS2 message)
Get RS1 rows (starts processing RS2 message)
Request next 500 rows from RS1 (blocked on RS2 + RS3)
Get RS2 rows (starts processing RS3 message)
Request next 500 rows from RS2 (blocked on RS3 + RS1)
Get RS3 rows (starts processing RS1 message)
=> RS3 had < 500 rows so it's done and returns first, even though the others are still processing

And this is where we end up with results from RS3 showing up "first".

In summary : 

1. The root problem here is that we've supported result set streaming for many years now, but it was basically built on a bunch of assumptions of how the tools service handles messages and returns results. The "best" fix here would be to guarantee that in our core Notebook code to guarantee order no matter how the messages are processed.
2. The above PR broke this assumption so we ended up with the above bug. The easiest solution would likely be to allow callers to getQueryRows to decide whether they want to paginate the data or not (basically letting Notebooks revert back to the previous functionality)

**Comment by Charles-Gagnon** [comment]
Confirmed this is happening for SQL kernels in Notebooks too. Although there is an additional layer of complexity there - because we have parallel message processing enabled for the subset calls (https://github.com/microsoft/sqltoolsservice/blob/main/src/Microsoft.SqlTools.ServiceLayer/QueryExecution/QueryExecutionService.cs#L180) then even the fix in my PR doesn't fix it. The root cause is the same (not handling out-of-order batches when recombining them), but it's happening even with the old way. I confirmed this happens even in older versions (ever since we enabled parallel processing in https://github.com/microsoft/sqltoolsservice/pull/1741)

So I'm going to leave this issue opened so we can track fully fixing it...

**Comment by Charles-Gagnon** [comment]
@cowhitley The fix has been released in our [insiders](https://github.com/Microsoft/azuredatastudio#try-out-the-latest-insiders-build-from-main-branch) build. Let us know if you still see any issues if you get a chance to try it out!

**Comment by Charles-Gagnon** [comment]
Moving to Nov since this only occurs when parallel processing is enabled (which is disabled by default outside of dev and marked as experimental).

@cheenamalhotra FYI, we won't be able to have query execution stuff (at least the SubsetRequest) be ran async unless this issue is fixed.

**Comment by Charles-Gagnon** [comment]
Enabled parallel message processing and couldn't repro issue so I'm closing this as resolved
